### PR TITLE
picom-allusive: init for 0.3.1

### PR DESCRIPTION
**Pull request by allusive-dev**
## Description of changes

Add picom-allusive: https://github.com/allusive-dev/picom-allusive
Add `allusive` to maintainers list

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by cafkafk — COMMENTED** [review]
refs: #257979

**Review by cafkafk — APPROVED** [review]
(see https://github.com/NixOS/nixpkgs/pull/257979#pullrequestreview-1652421984 as well)

File: pkgs/applications/window-managers/picom/picom-allusive.nix
Comment (NotAShelf) [inline pkgs/applications/window-managers/picom/picom-allusive.nix:18]: minor nitpick, but nixpkgs is trying to drop the redundant usage of `with lib;` and this could be replaced with lib.licenses and lib.maintainers

**Review by NotAShelf — APPROVED** [review]

File: pkgs/applications/window-managers/picom/picom-allusive.nix
Comment (allusive-dev) [reply pkgs/applications/window-managers/picom/picom-allusive.nix:18]: I tried using lib.licences before but that resulted in errors.

Comment (allusive-dev) [reply pkgs/applications/window-managers/picom/picom-allusive.nix]: Ok, i made the changes and pushed again.

**Review by NotAShelf — APPROVED** [review]
LGTM

**Review by cafkafk — APPROVED** [review]

**Comment by allusive-dev** [comment]
Just wondering is this waiting for more approvals or waiting to be merged?

File: pkgs/applications/window-managers/picom/picom-allusive.nix
Comment (samueldr) [inline pkgs/applications/window-managers/picom/picom-allusive.nix:16]: To prevent `postInstall` being removed from the overridden package breaking the eval:

```suggestion
  '' + (lib.optionalString (oldAttrs ? postInstall) oldAttrs.postInstall);
```

I would usually prefer/use a set-pattern argument (`{ postInstall ? "", ... }: {/*...*/}`) for this purpose, but `rec` is wrecking (heh) the scope here, meaning `postInstall` would infrec on the attrset's own `postInstall`.

Comment (allusive-dev) [reply pkgs/applications/window-managers/picom/picom-allusive.nix:16]: I have no idea what you are saying. Can you please explain it more simply?

Comment (samueldr) [reply pkgs/applications/window-managers/picom/picom-allusive.nix:16]: Sure!

* * *

### About `overrideAttrs`

The `overrideAttrs` function takes one argument; that argument is a function, which receives an *attribute set* as a parameter; the attribute set (if you allow the oversimplification) is all attributes provided to `mkDerivation` in the original derivation.

Here's how it looks like with `builtins.trace` in the repl, doing a "no-op" to `hello`.

```
nix-repl> hello.overrideAttrs (x: builtins.trace x x)
trace: { doCheck = true; meta = <CODE>; passthru = <CODE>; pname = "hello"; src = <CODE>; version = "2.12.1"; }
«derivation /nix/store/vz9naq2p06k5sagqz5mxcbmff0a4d3hb-hello-2.12.1.drv»
```

 

In your particular case, `oldAttrs` is the argument name to the function passed to `overrideAttrs`:

```text
picom.overrideAttrs (oldAttrs: rec {
                     ^^^^^^^^ This is the argument to the function given to `overrideAttrs`
```

And in turn, when referring to `oldAttrs.installPhase`, it refers to this attribute:

 - https://github.com/NixOS/nixpkgs/blob/5a9c737c587d2c34d63c5b3cb53c6ab0705bdf4f/pkgs/applications/window-managers/picom/default.nix#L89-L94

*if* this attribute ends-up being removed through a refactor of the `picom` package, the evaluation will fail; here's how it would fail, using `hello`, which doesn't have a `postInstall`:

```
nix-repl> hello.overrideAttrs (oldAttrs: { postInstall = "" + oldAttrs.postInstall; })
       error: attribute 'postInstall' missing

       at «string»:1:53:

            1| hello.overrideAttrs (oldAttrs: { postInstall = "" + oldAttrs.postInstall; })
             |                                                     ^
```

We can guard against this problem like so:

```
nix-repl> hello.overrideAttrs (oldAttrs: { postInstall = "" + (lib.optionalString (oldAttrs ? postInstall) oldAttrs.postInstall); })  
«derivation /nix/store/bvw2jx898i33dzih5zn4bwh499d7vns7-hello-2.12.1.drv»
```

Please keep on reading :)

* * *

### `a ? x` ? ??

 - https://nixos.org/manual/nix/stable/language/operators#has-attribute

Using the `?` operator we're checking the attribute set has the given attribute.

```
nix-repl> { a = 1; } ? a
true

nix-repl> { b = 1; } ? a 
false
```

* * *

### A set-pattern what now?

Second list item of the *Functions* part of the language constructs page:

 - https://nixos.org/manual/nix/stable/language/constructs#functions

A *set pattern argument* is a special kind of argument to a function that is close to what *destructuring* is in other languages.

Now, using our contrived example with `hello` in mind, we can instead do this:

```
nix-repl> hello.overrideAttrs ({ postInstall ? "", ... }: { postInstall = "" + postInstall; })                                       
«derivation /nix/store/bvw2jx898i33dzih5zn4bwh499d7vns7-hello-2.12.1.drv»
```

*But why can't we do this here?* As I alluded to, it's `rec` w`rec`king our plans.

```
nix-repl> hello.overrideAttrs ({ postInstall ? "", ... }: rec { postInstall = "" + postInstall; })
error:
...
       error: infinite recursion encountered

       at «string»:1:74:

            1| hello.overrideAttrs ({ postInstall ? "", ... }: rec { postInstall = "" + postInstall; })
             |                                                                          ^
```

`rec` makes the attribute set [a recursive set](https://nixos.org/manual/nix/stable/language/constructs#recursive-sets), which in turn, means that the attribute set's *own* attributes are now in scope. So the bare `postInstall` in the attribute set refers to the value it is trying to set, causing the infinite recursion.

In fact, this is entirely independent to the `overrideAttrs` function call!

```
nix-repl> rec { postInstall = "" + postInstall; }                                                  
{ postInstall = error: infinite recursion encountered

       at «string»:1:26:

            1| rec { postInstall = "" + postInstall; }
             |                          ^
```

* * *

Does that help you understand what I meant? Don't worry if you have more questions, do ask.

Comment (allusive-dev) [reply pkgs/applications/window-managers/picom/picom-allusive.nix:16]: Seems quite complex but I think I understand. I'll give it a try now.

**Comment by allusive-dev** [comment]
Hey @samueldr Thank you very much for the help and for merging my package!
Just wanted to know how long it will take before it is visible on `nixos.org` or usable in `configuration.nix`?

**Comment by cafkafk** [comment]
you can follow the package's progress here https://nixpk.gs/pr-tracker.html?pr=258711

**Comment by allusive-dev** [comment]
> you can follow the package's progress here https://nixpk.gs/pr-tracker.html?pr=258711

Thank you very much. I notice that doesn't have a label for the stable package. How will i know when it is available on 23.05?

**Comment by samueldr** [comment]
We don't usually backport new packages in the stable release.

Though with 23.11 upcoming, it won't be too long until a stable release has it in.

**Comment by allusive-dev** [comment]
> We don't usually backport new packages in the stable release.
> 
> Though with 23.11 upcoming, it won't be too long until a stable release has it in.

Awesome, thank you for letting me know.

**Pull request merged**
